azpubsub/contextvalidator/src/main/scala/kafka/security/auth/AzPubSubAclAuthorizer.scala

```diff
@@ -0,0 +1,459 @@
+
+              return false
+            }
+
+            LOGGER.info(s"Token is valid. ValidFrom: ${token("ValidFrom")}, ValidTo: ${token("ValidTo")}")
+
+            token("Roles").asJsonArray.iterator.map(_.to[String]).foreach(r => {
+              LOGGER.debug(s"Claim from json token: ${r}")
+              val prin = new KafkaPrincipal(KafkaPrincipal.ROLE_TYPE, r)
+
+              if(aclMatch(operation, resource, prin, session.clientAddress.getHostAddress, Allow, acls)){
+
+                LOGGER.info(s"Authorization for ${prin} operation ${operation} on resource ${resource} succeeded.")
+
+                newTimer(AzPubSubAclAuthorizer.TopicAuthorizationUsingTokenSuccessfulRateMs, TimeUnit.MILLISECONDS, TimeUnit.SECONDS)
+
+                return true
+              }
+            })
+
+            LOGGER.warn(s"The token doesn't have any role permitted to access the particular topic ${resource.name}.")
+
+            newTimer(AzPubSubAclAuthorizer.TokenNotAuthorizedForTopicRateMs, TimeUnit.MILLISECONDS, TimeUnit.SECONDS)
+
+            return false
+          }
+          case _ => {
+            LOGGER.warn(s"unknown principal rejected: ${session.principal}, accessing resource: ${resource}, operation: ${operation}")
+
+            return false
+          }
+        }
+      }
+      case _ => {
+
+        /**
+          * If the client is trying to access other resources like "Cluster/Group/DelegationToken/TxnId",
+          * This kind of requests should be always coming from brokers of the current cluster; otherwise, the request should be rejected.
+          */
+
+        LOGGER.debug(s"session client address: ${session.clientAddress.getHostAddress}")
+
+        if(!brokerHosts.contains(session.clientAddress.getHostAddress)) {
+
+          val allBrokers = zkClient.getAllBrokersInCluster
+
+          allBrokers.foreach(b => b.endPoints.foreach(e => {
+
+            brokerHosts += InetAddress.getByName(e.host).getHostAddress
+
+            LOGGER.debug(s"Kafka broker host : ${e.host}")
+          }))
+        }
+
+        if(!brokerHosts.contains (session.clientAddress.getHostAddress) ){
+
+          LOGGER.warn(s"Client is not broker and accessing ${resource.resourceType} rejected: ${session.clientAddress.getHostAddress}")
+
+          return false
+        }
+
+        LOGGER.debug(s"Client is broker and accessing ${resource.resourceType} allowed: ${session.clientAddress.getHostAddress}")
+
+        return true
+      }
+    }
+  }
+
+  protected def aclMatch(operations: Operation, resource: Resource, principal: KafkaPrincipal, host: String, permissionType: PermissionType, acls: Set[Acl]): Boolean = {
+
+    acls.find { acl =>
+
+      acl.permissionType == permissionType &&
+        (acl.principal == principal
+          || (principal.getPrincipalType == KafkaPrincipal.USER_TYPE && acl.principal == KafkaPrincipal.WildCardUserTypePrincipal)
```

Comment:
refactor this, use a boolean to get the value of the long expression, and make the conditions operated only on booleans

Reply:
I don't I need to. this is ACL matching logic, which is quite straight forward.